src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+{
+  public:
+    explicit NodeSetBasicNodeIds(Selection::Values&& values)
+        : values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */,
+                          const NodePopulation& /* unused */) const final {
+        return Selection::fromValues(values_.begin(), values_.end());
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", values_);
+    }
+
+  private:
+    Selection::Values values_;
+};
+
+using CompoundTargets = std::vector<std::string>;
+class NodeSetCompoundRule: public NodeSetRule
+{
+  public:
+    NodeSetCompoundRule(std::string attribute, const CompoundTargets& targets)
+        : attribute_(std::move(attribute))
+        , targets_(targets) {}
+
+    Selection materialize(const NodeSets& ns, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& target : targets_) {
+            ret = ret | ns.materialize(target, np);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", targets_);
+    }
+
+  private:
+    std::string attribute_;
+    CompoundTargets targets_;
+};
+
+int64_t get_integer_or_throw(const json& el) {
+    auto v = el.get<double>();
+    if (std::floor(v) != v) {
+        throw SonataError("Only allowed integers in node set rules");
+    }
+    return static_cast<int64_t>(v);
+}
+
+NodeSetRules _dispatch_node(const json& contents) {
+    NodeSetRules ret;
+
+    for (auto& el : contents.items()) {
+        const std::string& attribute = el.key();
+        if (el.value().is_number()) {
+            if (attribute == "population") {
+                throw SonataError("'population' must be a string");
+            }
+
+            int64_t v = get_integer_or_throw(el.value());
+            if (attribute == "node_id") {
+                if (v < 0) {
+                    throw SonataError("'node_id' must be positive");
+                }
+
+                ret.emplace_back(new NodeSetBasicNodeIds({static_cast<Selection::Value>(v)}));
+            } else {
+                ret.emplace_back(new NodeSetBasicRule<int64_t>(attribute, {v}));
+            }
+        } else if (el.value().is_string()) {
+            if (attribute == "node_id") {
+                throw SonataError("'node_id' must be numeric or a list of numbers");
+            }
+
+            if (attribute == "population") {
+                ret.emplace_back(new NodeSetBasicPopulation(el.value().get<std::string>()));
+            } else {
+                ret.emplace_back(
+                    new NodeSetBasicRule<std::string>(attribute, {el.value().get<std::string>()}));
+            }
+        } else if (el.value().is_array()) {
+            const auto array = el.value();
+
+            if (attribute == "population" && (array.size() != 1 || !array[0].is_string())) {
+                throw SonataError("'population' must be a single string");
+            }
```

Comment:
see my comment above #L58

Reply:
I think that comment got lost?

Reply:
This is the comment that have been killed during the UI madness ... 
It is part of a second round of comments. Put it here again : 

To me there is no reason why you cannot have a list of populations with this :
```
{
"pop1" : {"population" : "population1"},
"pop12" : {"population" : ["population1", "population2"]},
}
```
Allowing this would also allow to :
```
{
"pop1" : {"population" : ["population1"]},
}
```
without raising.

Reply:
ok, that's supported now.